File: android/app/src/main/java/com/gradethread/app/inventory/SizeBandsService.kt
package com.gradethread.app.inventory

import com.gradethread.app.platform.net.EdgeApi
import kotlinx.serialization.json.Json
import javax.inject.Inject
import javax.inject.Named
import javax.inject.Singleton

/**
 * US-2921: `GET /api/flipdesk/size-bands`.
 *
 * Fetches the expected-size band table for one brand + garment + department.
 * Kept out of [SizeCheck] on purpose: the check itself is pure arithmetic that
 * runs on every keystroke, and the only reason it needs a network at all is to
 * get the table once. The edge marks the response cacheable for half an hour and
 * it does not depend on the caller, so a stale table is never wrong, only old.
 *
 * A failure is NOT an error state. The size check is an assist; when the table
 * cannot be fetched the canvas shows no note and the seller carries on, which is
 * exactly what happens for a brand with no chart on file.
 */
@Singleton
class SizeBandsService @Inject constructor(
    /** The short-idle profile: this is a cached reference read, not a model call. */
    @Named("shared") private val edge: EdgeApi,
) {

    companion object {
        const val BANDS_PATH = "/api/flipdesk/size-bands"

        /** Half an hour, matching the endpoint's own Cache-Control. */
        const val CACHE_TTL_MILLIS = 30L * 60L * 1000L
    }

    suspend fun load(
        brand: String?,
        garment: String?,
        gender: String?,
    ): SizeCheck.BandsResponse {
        val garmentValue = garment?.trim().orEmpty()
        if (garmentValue.isEmpty()) return SizeCheck.BandsResponse.EMPTY

        val query = buildMap {
            put("garment", garmentValue)
            brand?.trim()?.takeIf { it.isNotEmpty() }?.let { put("brand", it) }
            gender?.trim()?.takeIf { it.isNotEmpty() }?.let { put("gender", it) }
        }

        return runCatching {
            val raw = edge.getRaw(BANDS_PATH, query, CACHE_TTL_MILLIS)
            json.decodeFromString(SizeCheck.BandsResponse.serializer(), raw)
        }.getOrDefault(SizeCheck.BandsResponse.EMPTY)
    }

    private val json = Json { ignoreUnknownKeys = true; isLenient = true }
}

File: ios/GradeThread/Inventory/ItemCanvas/SizeBandsService.swift
import Foundation

/// US-2920: fetch the expected-size band table for one brand + garment.
///
/// Kept out of `SizeCheck` on purpose: the check itself is pure arithmetic that
/// has to run on every keystroke, and the only reason it needs a network at all
/// is to get the table once. The edge marks the response cacheable for half an
/// hour and it does not depend on the caller, so a stale table is never wrong,
/// only old.
///
/// A failure is NOT an error state. The size check is an assist; when the table
/// cannot be fetched the canvas shows no note and the seller carries on, which
/// is exactly what happens for a brand with no chart on file.
enum SizeBandsService {
    /// Half an hour, matching the endpoint's own Cache-Control.
    static let cacheTTL: TimeInterval = 30 * 60

    static func load(
        brand: String?,
        garment: String?,
        gender: String?,
        api: EdgeAPI = .shared
    ) async -> SizeCheck.BandsResponse {
        let garmentValue = (garment ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        guard !garmentValue.isEmpty else { return .empty }

        var query = [URLQueryItem(name: "garment", value: garmentValue)]
        let brandValue = (brand ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        if !brandValue.isEmpty {
            query.append(URLQueryItem(name: "brand", value: brandValue))
        }
        let genderValue = (gender ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        if !genderValue.isEmpty {
            query.append(URLQueryItem(name: "gender", value: genderValue))
        }

        do {
            let response: SizeCheck.BandsResponse = try await api.getJSON(
                "/api/flipdesk/size-bands",
                query: query,
                cacheTTL: cacheTTL
            )
            return response
        } catch {
            return .empty
        }
    }
}
